### ISSUE TITLE
Error creating automation config secret: could not ensure scram credentials

### ISSUE THREAD

**Issue by Parsonswy** [mongodb/mongodb-kubernetes-operator#252]
Hello.
I am attempting to deploy a MongoDB instance using the community Kubernetes operator and have encountered an issue with the recent SCRAM changes.

I'm seeing this error in the operator logs and pasted my MongoDB configuration below.
```
2020-11-20T00:38:58.440Z        DEBUG   mongodb/replica_set_controller.go:184   Validating MongoDB.Spec {"ReplicaSet": "ncsa-insights/ncsa-insights-mongodb"}
2020-11-20T00:38:58.440Z        DEBUG   mongodb/replica_set_controller.go:193   Ensuring Automation Config for deployment       {"ReplicaSet": "ncsa-insights/ncsa-insights-mongodb"}
2020-11-20T00:38:58.473Z        DEBUG   scram/scram.go:100      password secret was not found, reading from credentials from secret/ncsa-insights-mongodb-admin.ncsa.tech-scram-operator-scram-credentials
2020-11-20T00:38:58.474Z        ERROR   mongodb/mongodb_status_options.go:83    error creating automation config secret: could not ensure scram credentials: Secret "ncsa-insights-mongodb-admin.ncsa.tech-scram-operator-scram-credentials" not found
github.com/mongodb/mongodb-kubernetes-operator/pkg/controller/mongodb.messageOption.ApplyOption
        /go/pkg/controller/mongodb/mongodb_status_options.go:83
github.com/mongodb/mongodb-kubernetes-operator/pkg/util/status.Update
        /go/pkg/util/status/status.go:25
github.com/mongodb/mongodb-kubernetes-operator/pkg/controller/mongodb.(*ReplicaSetReconciler).Reconcile
        /go/pkg/controller/mongodb/replica_set_controller.go:195
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        /root/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:256
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /root/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:232
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker
        /root/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:211
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1
        /root/go/pkg/mod/k8s.io/apimachinery@v0.17.5/pkg/util/wait/wait.go:152
k8s.io/apimachinery/pkg/util/wait.JitterUntil
        /root/go/pkg/mod/k8s.io/apimachinery@v0.17.5/pkg/util/wait/wait.go:153
k8s.io/apimachinery/pkg/util/wait.Until
        /root/go/pkg/mod/k8s.io/apimachinery@v0.17.5/pkg/util/wait/wait.go:88
```
```yaml
apiVersion: mongodb.com/v1
kind: MongoDB
metadata:
  name: ncsa-insights-mongodb
  namespace: ncsa-insights
spec:
  clusterName: ncsa-insights-mongodb
  type: ReplicaSet
  version: "4.2.7"
  persistent: true
  podSpec:
    persistence:
      single:
        storage: 40Gi
        storageClass: ncsa-nfs
  security:
    authentication:
      enabled: false
      modes:
        - SCRAM
  users:
    - name: "insights-admin"
      passwordSecretRef:
        key: "password"
        name: ncsa-insights-mongodb-admin
      roles:
        - db: "admin"
          name: "clusterAdmin"
        - db: "admin"
          name: "userAdminAnyDatabase"
        - db: "admin"
          name: "readWrite"
        - db: "admin"
          name: "userAdminAnyDatabase"
      scramCredentialsSecretName: ncsa-insights-mongodb-admin.ncsa.tech-scram-operator  #Read below, I know this is not what this value should be set to.
```
The [documentation says](https://github.com/mongodb/mongodb-kubernetes-operator/blob/e95497e6aec65617c5438d369bb22df9230ea1ff/docs/users.md) the `scramCredentialsSecretName` field should be created by the operator. I thought meant when I installed the operator a new secret would be created in that namespace and I would need to put that value in the field, but no applicable secret was created. The only secret in the store is the service account token, `mongodb-kubernetes-operator-token-b2wg4`. I tried providing my own name (above), but I expect that is not right based on what the documentation says.

Is there another step in the installation process for generating the SCRAM secret? I followed the [deployment documentation](https://github.com/mongodb/mongodb-kubernetes-operator/blob/e95497e6aec65617c5438d369bb22df9230ea1ff/docs/deploy-configure.md) and have looked through the [creating users documentation](https://github.com/mongodb/mongodb-kubernetes-operator/blob/e95497e6aec65617c5438d369bb22df9230ea1ff/docs/users.md) and don't see where that comes up.

Thank you.

**Comment by ciwarren** [comment]
Same Issue 👍

**Comment by NetworkGizmo** [comment]
I DEFINITELY have the same issue...

**Comment by bznein** [comment]
Hi @Parsonswy , @ciwarren  and @NetworkGizmo,

Thanks for your report, let me clarify the procedure (we will work on a better documentation, as it seems that it is not clear enough).

There are two different `secrets` involved in the procedure.
One is generated by the user: https://github.com/mongodb/mongodb-kubernetes-operator/blob/e95497e6aec65617c5438d369bb22df9230ea1ff/docs/users.md#create-a-user-secret and it contains the password that you want your user to have. This is what gets specified in `spec.users.passwordSecretRef.name`.


The other one is generated by the Operator, using the name provided in `spec.users.scramCredentialsSecretName`.

In @Parsonswy's example, what happens is that the secret referenced in `spec.users.passwordSecretRef.name` doesn't exist (that's safe to delete AFTER the operator has created it's own secret), so the Operator tried to read from the one he should have created automatically, but that doesn't exist either. My guess is that `ncsa-insights-mongodb-admin` was never manually created and so the operator has no way to create the user as it doesn't know what password it should have.

You could try creating a secret with this format:
```
apiVersion: v1
kind: Secret
metadata:
  name: ncsa-insights-mongodb-admin
type: Opaque
stringData:
  password: <password-for-your-user>
```

Please let me know if this helps

**Comment by Parsonswy** [comment]
Thank you for the quick response, @bznein.
Sorry, to waste you're time. I was creating the secret, but in one location I had named it `ncsa-insights-mongodb-admin` and the other `ncsa-insights-mongo-admin`.

Apart from #251, the only thing I found confusing in the docs was the [`scramCredentialsSecretName` explanation](https://github.com/mongodb/mongodb-kubernetes-operator/blob/master/docs/users.md). I personally read "the secret object being created by the operator" and thought it was something that existed as part of the operator deployment rather than something that was created after the Mongo deployment. In hindsight I am not sure what I concluded such a secret store would hold.

Thank you again.
